ch5/exercise5-9/expand.go

```diff
@@ -0,0 +1,43 @@
+package expand
+
+import (
+	"bytes"
+	"os"
+	"strings"
+)
+
+var f = func(x string) string {
+
+	args := os.Args[0:]
+
+	if len(args) == 0 {
+		return x
+	}
+	
+	var buf bytes.Buffer
+
+	for i := 0; i < len(args); i++ {
+		buf.WriteByte(args[i][0])
+	}
+
+	return x + buf.String()
+}
+
```

Comment:
Is this^ what you meant by:

> declare a variable of your function type and have this set based on an argument or flag.

Reply:
Kinda.  I like what you're doing with a dynamic number of repeats based on the number of cli args.  However, doing inline processing of `os.Args` is an antipattern  -- it is mixing business logic with IO.  Notice how in your test code, you have to mangle the real `os.Args` to properly test this code.  My recommendation always try to keep io logic close to main, and keep function like this more pure.  For example:

```go
import (
    "log"
    "os"
)

type expandFn func(s string) string

func expand(s string, f expandFn) string {
    // snip...
    return s
}

func double(s string) string {
    return s + s 
}

func triple(s string) string {
    return s + s + s 
}

func main() {
    var f expandFn
    prog := os.Args[0]
    requestedFn := os.Args[1] // ignoring error handling for brevity
    s := os.Args[2]           // ignoring error handling for brevity
    switch os.Args[1] {
    case "double":
        f = double
    case "triple":
        f = triple
    default:
        log.Fatalln("%s: unknown function: %s", prog, requestedFn)
    }   
    out := expand(s, f)
    log.Println(out)
}
```

Or, if you wanted something to dynamically generate a number of repeats:

```go
func timesX(x uint) expandFn {
    f := func(s string) string {
        var buf bytes.Buffer
        for i := 0; i < x; i++ {
            buf.WriteByte(args[i][0])
        }
        return x + buf.String()
    }
    return f
}

// and then later ...
n := uint(strconv.AtoI(os.Args[1]))
out := expand(timesX(n), s)


```

Either of the above grapple with functions as types and data -- which is the biggest take away from this exercise.